### PR TITLE
Fix memory leak

### PR DESCRIPTION
**Pull request by prismofeverything**
Based on a report that a multi-gen simulation run took more memory to complete than before the complexation change, this lead to the discovery of a memory leak in this repository. Ultimately it turned out to be quite an adventure into the low level internals of numpy and python/C interaction.

I used valgrind to find the leak, which is a pretty amazing tool if anyone ever needs to debug memory issues again (works for segfaults too): http://valgrind.org/

As it turns out, if you want to return an array from C as a numpy array, you can't free that pointer until after numpy is done with it. But when is numpy done with it? We don't know. I took this to mean that numpy would free it afterwards, but this assumption turned out to be incorrect.

The fix is to create empty numpy arrays of the right shape from C and copy the values from the C array into it. Then I was able to free those pointers. There were a couple of other improvements (I found a bug in the reaction selection process) and additions to the tests. 

Happy to answer any questions, let me know.

### CODE REVIEW

File: arrow/obsidianmodule.c [hunk 1]
```diff
@@ -214,9 +214,16 @@ Obsidian_evolve(ObsidianObject *self, PyObject *args)
...
   // Create new python numpy arrays from the raw C results
-  PyObject *time_obj = PyArray_SimpleNewFromData(1, steps, NPY_DOUBLE, result.time);
-  PyObject *events_obj = PyArray_SimpleNewFromData(1, steps, NPY_INT64, result.events);
-  PyObject *outcome_obj = PyArray_SimpleNewFromData(1, substrates, NPY_INT64, result.outcome);
+  PyArrayObject *time_obj = (PyArrayObject *) PyArray_SimpleNew(1, steps, NPY_DOUBLE);
+  memcpy(PyArray_DATA(time_obj), result.time, (sizeof (double)) * result.steps);
+  PyArrayObject *events_obj = (PyArrayObject *) PyArray_SimpleNew(1, steps, NPY_INT64);
+  memcpy(PyArray_DATA(events_obj), result.events, (sizeof (long)) * result.steps);
+  PyArrayObject *outcome_obj = (PyArrayObject *) PyArray_SimpleNew(1, substrates, NPY_INT64);
+  memcpy(PyArray_DATA(outcome_obj), result.outcome, (sizeof (long)) * self->substrates_count);
```
Comment (1fish2) [inline arrow/obsidianmodule.c]: Great work debugging this! FYI Python 3 improves memory debugging features, e.g. `--with-valgrind` compile option and `faulthandler` module.

In retrospect, we should've thought of this issue since numpy can't know what deallocator to call or whether you pointed it to a portion of a larger memory node.

(FYI, it looks like one [can register a destructor](https://stackoverflow.com/a/52737023/1682419), but letting numpy handle the memory is simpler.)

Please make the code check for memory allocation failures.

Technically `result.events` (et al) should be declared as an array of `int64_t`, not `long` whose size varies with the C compiler, the target CPU architecture, and compiler flags.

It's better to use `sizeof(result.events[0])` (et al) so it can't get out of sync with the declaration.

Similarly, `steps` and `substrates` should be arrays of `npy_intp`, although in this case the compiler should complain if the element size is wrong.

File: arrow/obsidianmodule.c [hunk 2]
```diff
@@ -472,6 +477,11 @@ _invoke_obsidian(PyObject *self, PyObject *args) {
   Py_XDECREF(substrates_indexes_array);
   Py_XDECREF(substrates_array);
 
+  // if something went wrong throw an error
+  if (obsidian == NULL) {
+    return NULL;
+  }
```
Comment (1fish2) [inline arrow/obsidianmodule.c]: Good catch on calling deref before this exit, however now there's no need to conditionally return NULL before returning NULL unless "throw an error" is supposed be a TODO for explicitly raising an error. Will returning NULL make the Python C API raise an error?

File: arrow/test/test_arrow.py [hunk 3]
```diff
@@ -100,28 +101,33 @@ def load_state(filename):
...
+    system = make_system(stoichiometric_matrix, rates, random_seed=np.random.randint(2**31))
     result = system.evolve(duration, initial_state)
 
     time = np.concatenate([[0.0], result['time']])
     events = result['events']
     occurrences = result['occurrences']
     outcome = result['outcome']
 
+    print(outcome)
+    print(np.where(outcome < 0))
```
Comment (1fish2) [inline arrow/test/test_arrow.py]: Printing unlabeled values is not great. Should it instead assert something, directly or via `np.assert...`?

File: arrow/test/test_arrow.py [hunk 4]
```diff
@@ -100,28 +101,33 @@ def load_state(filename):
...
-    return stoichiometric_matrix, rates, initial_state, final_state
+    return (stoichiometric_matrix, rates, initial_state, final_state)
 
 def complexation_test(make_system):
-    stoichiometric_matrix, rates, initial_state, final_state = load_complexation()
+    stoichiometric_matrix, rates, initial_state, final_state = load_complexation(prefix='beginning')
     duration = 1
 
-    system = make_system(stoichiometric_matrix, rates)
+    system = make_system(stoichiometric_matrix, rates, random_seed=np.random.randint(2**31))
```
Comment (1fish2) [inline arrow/test/test_arrow.py]: Does this seed the random number generator? It's good to have repeatable unit tests.

File: arrow/test/test_arrow.py [hunk 5]
```diff
@@ -173,11 +179,38 @@ def test_compare_runtime():
...
+    for i in range(amplify):
+        memory = this.memory_info().rss
+        if (memory != memory_previous):
+            print('memory increase iteration {}: {}'.format(i, memory))
+            memory_previous = memory
+
+        result = system.evolve(duration, initial_state)
+        difference = np.abs(final_state - result['outcome']).sum()
+
+        print('difference is {}'.format(difference))
```
Comment (1fish2) [inline arrow/test/test_arrow.py]: The difference in what?

File: arrow/obsidian.c [hunk 6]
```diff
@@ -181,7 +181,7 @@ evolve(MTState *random_state,
 
     // If the total is zero, then we have no more reactions to perform and can exit
     // early
-    if (total == 0.0) {
+    if (!(total > 0.0)) {
```
[the remark this answers is not in the document]
Comment (prismofeverything) [reply arrow/obsidian.c]: There was a period while debugging this where I had some floating point numbers that read to `0` out to 50 decimal places, but still did not trigger this condition. I found some things online about the perils of comparing floating point numbers for equality so I left this in here. I guess I could go either way, but it seemed like those values were infinitesimally positive.

File: arrow/obsidianmodule.c (hunk 1 above)
Comment (prismofeverything) [reply arrow/obsidianmodule.c]: > In retrospect, we should've thought of this issue since numpy can't know what deallocator to call or whether you pointed it to a portion of a larger memory node.

Yeah you mentioned this exact section as well. I was going off of the big red warning box in the numpy C api docs here: https://docs.scipy.org/doc/numpy-1.15.0/reference/c-api.array.html, which were correct, but my assumption was not. Good find on the destructor! But yes, I agree this is simpler. 

> Technically `result.events` (et al) should be declared as an array of `int64_t`, not `long` whose size varies with the C compiler, the target CPU architecture, and compiler flags.

Okay, I'll give that a shot.

File: arrow/obsidianmodule.c (hunk 2 above)
Comment (prismofeverything) [reply arrow/obsidianmodule.c]: Yes, returning `NULL` from a C function indicates to the python api that there is an error. All of these python interfacing functions expect some kind of non-NULL `PyObject *`.

File: arrow/test/test_arrow.py (hunk 3 above)
Comment (prismofeverything) [reply arrow/test/test_arrow.py]: Ah, missed those, not intended for production.

File: arrow/test/test_arrow.py (hunk 4 above)
Comment (prismofeverything) [reply arrow/test/test_arrow.py]: Yep, that was for my own purposes during testing but I can take it out now.

File: arrow/obsidian.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (1fish2) [reply arrow/obsidian.c]: In that case it should exit when `total < threshold` (or  `total <= threshold`). `!(total > 0.0)` is the same as `total <= 0.0` except that all comparisons with NaN values return false. (There are several NaN values.) If it gets NaN values then you could go with `!(total > threshold)` plus a comment.

IEEE floats and doubles can't represent infinitesimals. The smallest positive is about 2**-1022. The computation could fail to converge on 0 if the deltas get truncated to 0.

[the remark this answers is not in the document]
Comment (prismofeverything) [reply arrow/obsidian.c]: Cool, I guess by `infinitesimal` I meant `practically infinitesimal` which is what I would consider 2**-1022 (is there an effort to make an actual computational implementation of mathematical infinitesimals?). I had trouble actually finding what the significant parts of a value that small was, I just knew it tested != 0. Granted I didn't investigate much further than that, thanks for all the background on this.

**Comment by jmason42** [comment]
Ignoring the alternative propensity calculations for the moment: It would take some refactoring but there is a way to do this exactly.  The total propensity is (obviously) the sum of all propensities, while each propensity is some float (a positive rate constant) times the number of ways in which molecule can possibly interact with each other, which is itself always a non-negative integer.  So, if you think about it, the only way for the overall propensity (a float) to be zero is if all possible interaction #'s for all reactions are also zero, which can be checked without requiring any arbitrary threshold.

Now, as far as the alternative propensity calculations go, things become more complicated.  There's no strict enforcement of what value the alternative propensities can take, so it could be nonzero even when a reaction really shouldn't happen (@prismofeverything is there handling for this?).  Furthermore there's no prior calculation of the number of possible interactions between molecules, so therefore there are no integer values to check for zeros.  So we're back to checking an arbitrary threshold.  *However*, as I mentioned, we ought to be making sure that there are enough molecules for one reaction to occur anyway, otherwise there's the risk that copy numbers go negative.  So if that logic was introduced - which would be as simple as checking that the number of reactant molecules of each type is at least as great as their stoichiometry - then, again, we can do exact comparisons, and no thresholding is necessary.

In short - the question isn't "Is the total propensity zero?", it's "Can any more reactions possibly occur?".  They're equivalent in a mathematical sense but not in a computational sense.

**Comment by prismofeverything** [comment]
Awesome, thanks for that rundown @jmason42. Based on these considerations and my subsequent testing I am going to leave the condition as is `if (total <= 0.0)` as we are currently only calculating propensities in one way, and as we add new forms we can revisit the considerations for what this threshold should be. 

The artifacts I was seeing (infinitesimals) were temporary issues based on some memory access issues I was having while debugging this, and I have not seen them since those were fixed. All of my tests currently reach a total of exactly `0.0`.

File: arrow/test/test_arrow.py (hunk 5 above)
Comment (prismofeverything) [reply arrow/test/test_arrow.py]: This reminds me of an old joke: 

"What is the difference between a duck?"
